src/dune_rules/foreign_rules.ml

```diff
@@ -75,19 +75,51 @@ let include_dir_flags ~expander ~dir (stubs : Foreign.Stubs.t) =
 
 let build_c ~kind ~sctx ~dir ~expander ~include_flags (loc, src, dst) =
   let ctx = Super_context.context sctx in
+  let project = Super_context.find_scope_by_dir sctx dir |> Scope.project in
   let flags =
     let ctx_flags =
       match kind with
-      | Foreign_language.C ->
+      | Foreign_language.C -> (
         let cfg = ctx.ocaml_config in
-        List.concat
-          [ Ocaml_config.ocamlc_cflags cfg
-          ; Ocaml_config.ocamlc_cppflags cfg
-          ; Fdo.c_flags ctx
-          ]
+        match Dune_project.future_c_and_cxx_flags_handling project with
+        | None
+        | Some false ->
+          (* In dune < 2.8 flags from ocamlc_config are always added *)
+          List.concat
+            [ Ocaml_config.ocamlc_cflags cfg
+            ; Ocaml_config.ocamlc_cppflags cfg
+            ; Fdo.c_flags ctx
+            ]
+        | Some true -> Fdo.c_flags ctx )
       | Foreign_language.Cxx -> Fdo.cxx_flags ctx
     in
     let flags = Foreign.Source.flags src in
+    (* DUNE3 will have [future_c_and_cxx_flags_handling] disabled by default. To
+       guide users toward this change we emit a warning when dune_lang is >=
+       1.8, [future_c_and_cxx_flags_handling] is not specified in the
+       [dune-project] file (thus defaulting to [true]) and the [:standard] set
+       of flags has been overriden *)
+    let has_standard =
+      Ordered_set_lang.Unexpanded.encode flags
+      |> List.map ~f:Dune_lang.to_string
+      |> List.exists ~f:(String.equal ":standard")
```

Comment:
Could you add an `has_standard` function that works directly on the ast? This is consistent with how we do `is_standard` as well.